File: swift/BCEnvelope/Tests/BCEnvelopeTests/SSHTests.swift
import Testing
import BCComponents
import BCEnvelope
import Foundation

struct SSHTests {
    @Test func testSSHSignedPlaintext() throws {
        let aliceSSHPrivateKey = try alicePrivateKeys
            .sshSigningPrivateKey(.ed25519, comment: "alice@example.com")
        let aliceSSHPublicKey = try aliceSSHPrivateKey.publicKey()

        let options = SigningOptions.ssh(namespace: "test", hashAlg: .sha256)
        let envelope = try Envelope(plaintextHello)
            .addSignatureOpt(aliceSSHPrivateKey, options: options, metadata: nil)
            .checkEncoding()
        let ur = envelope.ur

        #expect(envelope.format() ==
        """
        "Hello." [
            'signed': Signature
        ]
        """)

        let receivedEnvelope = try Envelope(ur: ur).checkEncoding()
        let receivedPlaintext = try receivedEnvelope
            .verifySignatureFrom(aliceSSHPublicKey)
            .extractSubject(String.self)
        #expect(receivedPlaintext == plaintextHello)

        let carolSSHPublicKey = try carolPrivateKeys
            .sshSigningPrivateKey(.ed25519, comment: "carol@example.com")
            .publicKey()

        #expect(throws: (any Swift.Error).self) {
            try receivedEnvelope.verifySignatureFrom(carolSSHPublicKey)
        }

        let verifiers: [any Verifier] = [aliceSSHPublicKey, carolSSHPublicKey]
        try receivedEnvelope.verifySignaturesFromThreshold(verifiers, threshold: 1)

        #expect(throws: (any Swift.Error).self) {
            try receivedEnvelope.verifySignaturesFromThreshold(verifiers, threshold: 2)
        }
    }
}

File: csharp/BCEnvelope/BCEnvelope.Tests/SshTests.cs
using BlockchainCommons.BCComponents;
using BlockchainCommons.BCEnvelope;

namespace BlockchainCommons.BCEnvelope.Tests;

public sealed class SshTests
{
    [Fact]
    public void TestSshSignedPlaintext()
    {
        TagsRegistry.RegisterTags();

        // Generate SSH Ed25519 keypair
        var (aliceSshPrivateKey, aliceSshPublicKey) =
            SignatureScheme.SshEd25519.KeypairOpt("alice@example.com");
        var (carolSshPrivateKey, carolSshPublicKey) =
            SignatureScheme.SshEd25519.KeypairOpt("carol@example.com");

        // Alice sends a signed plaintext message to Bob.
        var options = new SigningOptions.SshOptions("test", "sha256");
        var envelope = TestData.HelloEnvelope()
            .AddSignatureOpt(aliceSshPrivateKey, options, null)
            .CheckEncoding();
        var ur = BCUR.UR.Create("envelope", envelope.TaggedCbor());

        var expectedFormat =
            "\"Hello.\" [\n" +
            "    'signed': Signature(SshEd25519)\n" +
            "]";
        Assert.Equal(expectedFormat, envelope.Format());

        // Bob receives the envelope.
        var receivedUr = BCUR.UR.FromUrString(ur.ToUrString());
        var receivedEnvelope = Envelope.FromTaggedCbor(receivedUr.Cbor)
            .CheckEncoding();

        // Bob receives the message, validates Alice's signature, and reads
        // the message.
        var verifiedEnvelope = receivedEnvelope
            .VerifySignatureFrom(aliceSshPublicKey);
        var receivedPlaintext = verifiedEnvelope.ExtractSubject<string>();
        Assert.Equal("Hello.", receivedPlaintext);

        // Confirm that it wasn't signed by Carol.
        Assert.Throws<EnvelopeException>(() =>
            receivedEnvelope.VerifySignatureFrom(carolSshPublicKey));

        // Confirm that it was signed by Alice OR Carol.
        receivedEnvelope.VerifySignaturesFromThreshold(
            new IVerifier[] { aliceSshPublicKey, carolSshPublicKey },
            1);

        // Confirm that it was not signed by Alice AND Carol.
        Assert.Throws<EnvelopeException>(() =>
            receivedEnvelope.VerifySignaturesFromThreshold(
                new IVerifier[] { aliceSshPublicKey, carolSshPublicKey },
                2));
    }
}
